src/ScheduleFunctions.cpp

```diff
@@ -390,305 +394,282 @@ Stmt build_provide_loop_nest(const map<string, Function> &env,
-            vector<Expr> cleanup_args;
-
-            // Make a struct with the buffers and their uncropped parents
-            for (auto p : cropped_buffers) {
-                // The cropped buffer_t
-                cleanup_args.push_back(p.first);
-                // Its parent
-                cleanup_args.push_back(p.second);
+        for (const auto &buffer: buffers_contents_to_annotate) {
+            // Return type is really 'void', but no way to represent that in our IR.
+            // Precedent (from halide_print, etc) is to use Int(32) and ignore the result.
+            Stmt mark_contents = Evaluate::make(
+                    Call::make(Int(32), "halide_msan_annotate_buffer_is_initialized", {buffer}, Call::Extern));
+            if (!is_no_op(annotate)) {
+                annotate = Block::make(annotate, mark_contents);
+            } else {
+                annotate = mark_contents;
             }
+        }
+    }
 
-            if (cropped_buffers.size() > 1) {
-                // NULL-terminate it
-                cleanup_args.push_back(make_zero(type_of<struct halide_buffer_t *>()));
-            }
+    // Make the extern call
+    Expr e = f.make_call_to_extern_definition(extern_call_args, target);
 
-            Expr cleanup_struct = Call::make(Handle(),
-                                             Call::make_struct,
-                                             cleanup_args,
-                                             Call::Intrinsic);
+    // Check if it succeeded
+    string result_name = unique_name('t');
+    Expr result = Variable::make(Int(32), result_name);
+    Expr error = Call::make(Int(32), "halide_error_extern_stage_failed",
+                            {extern_name, result}, Call::Extern);
+    Stmt check = AssertStmt::make(EQ::make(result, 0), error);
+    check = LetStmt::make(result_name, e, check);
 
-            // We have to anticipate failures in the extern stage, so
-            // call this by registering a destructor before the extern
-            // call and triggering it afterwards using an Allocate node.
-            string destructor_name = unique_name('d');
-            const char *fn = (cropped_buffers.size() == 1 ?
-                              "_halide_buffer_retire_crop_after_extern_stage" :
-                              "_halide_buffer_retire_crops_after_extern_stage");
-            check = Allocate::make(destructor_name, Handle(), MemoryType::Stack, {},
-                                   const_true(), check, cleanup_struct, fn);
-        }
+    if (!cropped_buffers.empty()) {
+        // We need to clean up the temporary crops we made for the
+        // outputs in case any of them have device allocations.
+        vector<Expr> cleanup_args;
 
-        for (size_t i = 0; i < lets.size(); i++) {
-            check = LetStmt::make(lets[i].first, lets[i].second, check);
+        // Make a struct with the buffers and their uncropped parents
+        for (auto p : cropped_buffers) {
```

Comment:
may as well make this const auto & while you're here

Reply:
Sure